src/hazelcore/dynamics/EvaluatorStep.re

```diff
@@ -0,0 +1,504 @@
+open Sexplib.Std;
+
+module EvalCtx = {
+  [@deriving sexp]
+  type t =
+    | Mark
+    | Ap1(t, DHExp.t)
+    | Ap2(DHExp.t, t)
+    | BinBoolOp1(DHExp.BinBoolOp.t, t, DHExp.t)
+    | BinBoolOp2(DHExp.BinBoolOp.t, DHExp.t, t)
+    | BinIntOp1(DHExp.BinIntOp.t, t, DHExp.t)
+    | BinIntOp2(DHExp.BinIntOp.t, DHExp.t, t)
+    | BinFloatOp1(DHExp.BinFloatOp.t, t, DHExp.t)
+    | BinFloatOp2(DHExp.BinFloatOp.t, DHExp.t, t)
+    | Cons1(t, DHExp.t)
+    | Cons2(DHExp.t, t)
+    | Pair1(t, DHExp.t)
+    | Pair2(DHExp.t, t)
+    | Let1(DHPat.t, t, DHExp.t)
+    | Let2(DHPat.t, DHExp.t, t)
+    | Inj(HTyp.t, InjSide.t, t)
+    | NonEmptyHole(
+        ErrStatus.HoleReason.t,
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+      )
+    | FixF(Var.t, HTyp.t, t)
+    | Cast(t, HTyp.t, HTyp.t)
+    | FailedCast(t, HTyp.t, HTyp.t)
+    | InvalidOperation(t, InvalidOperationError.t)
+    | ConsistentCase(t, list(DHExp.rule), int)
+    | InconsistentBranches(
+        MetaVar.t,
+        MetaVarInst.t,
+        VarMap.t_(DHExp.t),
+        t,
+        list(DHExp.rule),
+        int,
+      );
+};
+
+// Copy from Evaluator.re
+[@deriving sexp]
+type ground_cases =
+  | Hole
+  | Ground
+  | NotGroundOrHole(HTyp.t) /* the argument is the corresponding ground type */;
+
+let grounded_Arrow = NotGroundOrHole(Arrow(Hole, Hole));
+let grounded_Sum = NotGroundOrHole(Sum(Hole, Hole));
+let grounded_Prod = length =>
+  NotGroundOrHole(Prod(ListUtil.replicate(length, HTyp.Hole)));
+let grounded_List = NotGroundOrHole(List(Hole));
+
+let ground_cases_of = (ty: HTyp.t): ground_cases =>
+  switch (ty) {
+  | Hole => Hole
+  | Bool
+  | Int
+  | Float
+  | Arrow(Hole, Hole)
+  | Sum(Hole, Hole)
+  | List(Hole) => Ground
+  | Prod(tys) =>
+    if (List.for_all(HTyp.eq(HTyp.Hole), tys)) {
+      Ground;
+    } else {
+      tys |> List.length |> grounded_Prod;
+    }
+  | Arrow(_, _) => grounded_Arrow
+  | Sum(_, _) => grounded_Sum
+  | List(_) => grounded_List
+  };
+// End of copying from Evaluator.re
+
+let is_val = (d: DHExp.t): bool =>
+  switch (d) {
+  | BoundVar(_)
+  | Lam(_, _, _)
+  | ListNil(_)
+  | BoolLit(_)
+  | IntLit(_)
+  | FloatLit(_)
+  | Triv => true
+  | _ => false
+  };
+
+let rec is_boxedval = (d: DHExp.t): bool =>
+  switch (d) {
+  | Cast(d1, Arrow(ty1, ty2), Arrow(ty3, ty4)) =>
+    is_boxedval(d1) && !(HTyp.eq(ty1, ty3) && HTyp.eq(ty2, ty4))
+  | Cast(d1, ty, Hole) => ground_cases_of(ty) == Ground && is_boxedval(d1)
+  | _ => is_val(d)
```

Comment:
you'll need boxedval cases for other compound types as well (products, sum, lists). you can use this system called the Gradualizer which can generate the necessary rules:

http://www.cs.uml.edu/~mcimini/gradualizerDynamicSemantics/